src/inmanta/util/collections.py

```diff
@@ -0,0 +1,167 @@
+        return self._primary.get(key, default)
+
+    def get_reverse(self, key: S, default: Optional[Set[P]] = None) -> Optional[Set[P]]:
+        return self._reverse.get(key, default)
+
+    # Implement MutableMapping interface
+
+    def __getitem__(self, key: P) -> Set[S]:
+        return self._primary.__getitem__(key)
+
+    def __setitem__(self, key: P, value: Set[S]) -> None:
+        self.set_primary(key, value)
+
+    def __delitem__(self, key: P) -> None:
+        if key not in self._primary:
+            raise KeyError(key)
+        # trim reverse mapping
+        self.set_primary(key, set())
+        # delete from primary mapping
+        del self._primary[key]
+
+    def __iter__(self) -> Iterator[P]:
+        return iter(self._primary)
+
+    def __len__(self) -> int:
+        return len(self._primary)
+
+    def __repr__(self) -> str:
+        return f"BidirectionalManyToManyMapping({self._primary!r})"
+
+    def __str__(self) -> str:
+        return str(self._primary)
+
+
+class _BidirectionalManyToManyMappingReverse(BidirectionalManyToManyMapping[S, P], Generic[P, S]):
+    def __init__(self, base: BidirectionalManyToManyMapping[P, S]) -> None:
+        self._base = base
+        self._primary = self._base._reverse
+        self._reverse = self._base._primary
+
+    def reverse(self) -> BidirectionalManyToManyMapping[P, S]:
+        return self._base
+
+
+# TODO: drop or move to test case
+import math
+import random
+import time
+from collections.abc import Mapping
+from dataclasses import dataclass
+
+@dataclass
+class ScalingTest:
+    size: int
+    nb_targets: int
+    nb_updates: int
+
+    @staticmethod
+    def generate_mapping(*, size: int, nb_targets: int, target_range: Optional[int] = None) -> dict[int, set[int]]:
+        return {
+            i: {random.randint(1, target_range if target_range is not None else size) for _ in range(nb_targets)}
+            for i in range(size)
+        }
+
+    def test(self) -> None:
+        init_mapping: Mapping[int, Set[int]] = self.generate_mapping(size=self.size, nb_targets=self.nb_targets)
+        updates: Mapping[int, Set[int]] = self.generate_mapping(
+            size=self.nb_updates, nb_targets=self.nb_targets, target_range=self.size
+        )
+
+        m: BidirectionalManyToManyMapping[int, int] = BidirectionalManyToManyMapping()
+        time_start: float = time.process_time()
+        for k, v in init_mapping.items():
+            m[k] = v
+        time_populated: float = time.process_time()
+        for k, v in updates.items():
+            m[k] = v
+        time_updated: float = time.process_time()
+
+        print(
+            f"Initial population with {self.size} items with {self.nb_targets} each: {time_populated - time_start},"
+            f" update: of {self.nb_updates} items: {time_updated - time_populated}."
+        )
+
+
+ScalingTest(size=100, nb_targets=10, nb_updates=10).test()
```

Comment:
How does it perform?

Reply:
On my laptop:
```
Initial population with 100 items with 10 each: 0.00030825400000000086, update: of 10 items: 3.2813999999999066e-05.
Initial population with 100000 items with 10 each: 0.6959348740000001, update: of 100 items: 0.0007234259999999715.
Initial population with 100000 items with 10 each: 0.665464032, update: of 1000 items: 0.00753114900000007.
Initial population with 100000 items with 10 each: 0.6574984850000001, update: of 10000 items: 0.077292629.
Initial population with 100000 items with 10 each: 0.7974232150000002, update: of 100000 items: 0.8767623789999996.
Initial population with 1000000 items with 5 each: 6.596546615999998, update: of 1000 items: 0.005180475000003071.
Initial population with 1000000 items with 10 each: 10.285557354999998, update: of 1000 items: 0.009985014000001513.
Initial population with 10000 items with 10000 each: 16.487250724000006, update: of 10 items: 0.014592215999996938.
```

I think the second to last is representative of the scale we're aiming for? 1M resources, with 10 requires each (which is probably a lot for the average), in resources sets of 1000 each. First `put_version` might take 10s, but an update for a single `put_partial` should be snappy at `1ms`.

So (as can be expected) it scales pretty linearly with the size of the resources and the number of requires edges. The fact that it's linear is great, because it means that simply speeding up the raw efficiency (e.g. with Rust as you pointed out in another comment) can do a lot for us as our use case scales.